### PR TITLE
Implement deferred interaction triggers

### PR DESCRIPTION
**Pull request by crisbeto**
Includes a few prerequisites, as well as the actual implementation, for the `on interaction` and `prefetch on interaction` triggers. The changes here will be the base for implementing the `on viewport` and `on hover` triggers as well. Includes the following commits:

### refactor(compiler): extract deferred block trigger information
Reworks the compiler to use the API introduced in https://github.com/angular/angular/pull/51816 to match triggers to the element nodes they point to. This will be used to generate the new instructions for `on interaction` and `prefetch on interaction`.

### refactor(compiler): implement final instruction generation for interaction triggers
Updates the logic that generates the instructions for the `on interaction` and `prefetch on interaction` triggers to their final shape. Now the instructions take two arguments:
1. `triggerIndex` - index at which to find the trigger in the view where it will be rendered.
2. `walkUpTimes` - tells the runtime how many views up it needs to go to find the trigger element. If the argument is omitted, it means that the trigger is in the same view as the deferred block. A positive number means that the runtime needs to go up X amount of times to find the trigger. A negative number means that the trigger is inside the root view of the placeholder block. Negative numbers are capped at -1 since the placeholder is always in the same position at runtime.

### feat(core): implement deferred block interaction triggers
Adds the implementation for the `on interaction` and `prefetch on interaction` triggers.

### CODE REVIEW

File: packages/core/src/render3/instructions/defer_event_manager.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    for (const callback of this.callbacks) {
+      callback();
+    }
+  }
+}
+
+/** Manages events related to deferred blocks. */
+export class ɵDeferEventManager {
+  /** @nocollapse */
+  static ɵprov = /** @pureOrBreakMyCode */ ɵɵdefineInjectable({
```
Comment (crisbeto) [inline packages/core/src/render3/instructions/defer_event_manager.ts:31]: I made this class an injectable, because we'll need access to `NgZone` for some of the other triggers.

File: packages/core/src/render3/instructions/defer.ts [hunk 2]
```diff
@@ -33,7 +35,7 @@ import {ɵɵtemplate} from './template';
  * only placeholder content is rendered (if provided).
  */
 function shouldTriggerDeferBlock(injector: Injector): boolean {
-  const config = injector.get(DEFER_BLOCK_CONFIG, {optional: true});
+  const config = injector.get(DEFER_BLOCK_CONFIG, null, {optional: true});
```
Comment (crisbeto) [inline packages/core/src/render3/instructions/defer.ts:38]: Unrelated change, I just noticed that we were using a deprecated signature.

File: packages/core/test/acceptance/defer_spec.ts [hunk 3]
```diff
@@ -1009,4 +1009,471 @@ describe('#defer', () => {
...
+         fixture.detectChanges();
+         expect(fixture.nativeElement.textContent.trim()).toBe('Placeholder');
+
+         fixture.nativeElement.querySelector('button').click();
+         fixture.detectChanges();
+         flush();
+         expect(fixture.nativeElement.textContent.trim()).toBe('Main content');
+       }));
+
+    it('should resolve a trigger on a component outside the defer block', fakeAsync(() => {
```
Comment (crisbeto) [inline packages/core/test/acceptance/defer_spec.ts:1047]: I had to use `fakeAsync` here, because I couldn't get the `afterRender` to flush reliably in the browser tests.

File: packages/core/test/acceptance/defer_spec.ts [hunk 4]
```diff
@@ -1009,4 +1009,471 @@ describe('#defer', () => {
...
+
+         fixture.nativeElement.querySelector('button').click();
+         fixture.detectChanges();
+         flush();
+         expect(fixture.nativeElement.textContent.trim()).toBe('Main content');
+       }));
+  });
+
+  describe('interaction triggers', () => {
+    it('should load the deferred content when the trigger is clicked', fakeAsync(() => {
```
Comment (crisbeto) [inline packages/core/test/acceptance/defer_spec.ts:1235]: I had to use `fakeAsync` here, because I couldn't get the `afterRender` to fire predictably in the browser tests.

File: packages/core/src/render3/instructions/defer.ts [hunk 5]
```diff
@@ -262,6 +283,111 @@ export function ɵɵdeferPrefetchOnViewport(target?: unknown) {}  // TODO: imple
...
+    const triggerLView = getTriggerLView(initialLView, tNode, walkUpTimes);
+    const lDetails = getLDeferBlockDetails(initialLView, tNode);
+    const renderedState = lDetails[DEFER_BLOCK_STATE];
+
+    // If the block was loaded before the trigger was resolved, we don't need to do anything.
+    if (renderedState !== DeferBlockInternalState.Initial &&
+        renderedState !== DeferBlockState.Placeholder) {
+      afterRenderRef.destroy();
+      return;
+    }
```
Comment (AndrewKushnir) [inline packages/core/src/render3/instructions/defer.ts:355]: I'd propose moving this check up and call `getTriggerLView` after this check, so we don't lookup trigger LView in wrong (non-placeholder) views.

File: packages/core/src/render3/instructions/defer.ts [hunk 6]
```diff
@@ -262,6 +283,111 @@ export function ɵɵdeferPrefetchOnViewport(target?: unknown) {}  // TODO: imple
...
+  // A positive value or zero means that the trigger is in a parent view.
+  if (walkUpTimes >= 0) {
+    return walkUpViews(walkUpTimes, deferredHostLView);
+  }
+
+  // If the value is negative, it means that the trigger is inside the placeholder.
+  const deferredContainer = deferredHostLView[deferredTNode.index];
+  ngDevMode && assertLContainer(deferredContainer);
+  const triggerLView = deferredContainer[CONTAINER_HEADER_OFFSET] ?? null;
+
```
Comment (AndrewKushnir) [inline packages/core/src/render3/instructions/defer.ts:310]: I'd propose to also check that the LView in this container is indeed a placeholder one.

```
    if (ngDevMode) {
      const lDetails = getLDeferBlockDetails(deferredHostLView, deferredTNode);
      const renderedState = lDetails[DEFER_BLOCK_STATE];
      assertEqual(renderedState, DeferBlockState.Placeholder,
        'Expected a placeholder to be rendered in this defer block.');
    }
```

File: packages/core/src/render3/instructions/defer.ts [hunk 7]
```diff
@@ -262,6 +283,111 @@ export function ɵɵdeferPrefetchOnViewport(target?: unknown) {}  // TODO: imple
...
+  // We need to null check, because the placeholder might not have been rendered yet.
+  if (ngDevMode && triggerLView != null) {
```
Comment (AndrewKushnir) [inline packages/core/src/render3/instructions/defer.ts:312]:
```suggestion
  if (ngDevMode && triggerLView !== null) {
```

File: packages/core/src/render3/instructions/defer.ts [hunk 8]
```diff
@@ -262,6 +283,111 @@ export function ɵɵdeferPrefetchOnViewport(target?: unknown) {}  // TODO: imple
...
+    }
+
+    const element = getTriggerElement(triggerLView, triggerIndex);
+    const eventManager = injector.get(ɵDeferEventManager);
+    const cleanup = eventManager.onInteraction(element, () => {
+      callback();
+      removeLViewOnDestroy(triggerLView, cleanup);
+      if (initialLView !== triggerLView) {
+        removeLViewOnDestroy(initialLView, cleanup);
+      }
```
Comment (AndrewKushnir) [inline packages/core/src/render3/instructions/defer.ts:376]: 👍

File: packages/core/src/render3/instructions/defer_event_manager.ts [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+  listener = () => {
+    for (const callback of this.callbacks) {
+      callback();
+    }
+  }
+}
+
+/** Manages events related to deferred blocks. */
+export class ɵDeferEventManager {
```
Comment (AndrewKushnir) [inline packages/core/src/render3/instructions/defer_event_manager.ts:29]: Quick question:  do we plan to reuse this class for other triggers? I think `interaction` and `hover` cases are ok, but for the `viewport` case we'd need an intersection observer, so putting everything in one class would make the code non-tree-shakable.

File: packages/core/src/render3/instructions/defer_event_manager.ts [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+      // at the same time. Deferred blocks are unlikely to be used in such a way.
+      // 2. Matching events to their target isn't free. For each `click` and `keydown` event we
+      // would have look through all the triggers and check if the target either is the element
+      // itself or it's contained within the element. Given that `click` and `keydown` are some
+      // of the most common events, this may end up introducing a lot of runtime overhead.
+      // 3. We're still registering only two events per element, no matter how many deferred blocks
+      // are referencing it.
+      entry = new DeferEventManagerEntry();
+      trigger.addEventListener('click', entry.listener, EVENT_LISTENER_OPTIONS);
+      trigger.addEventListener('keydown', entry.listener, EVENT_LISTENER_OPTIONS);
```
Comment (AndrewKushnir) [inline packages/core/src/render3/instructions/defer_event_manager.ts:62]: I'd propose moving events to a separate array, so we can run `removeEventListener` using the same list.

File: packages/core/test/acceptance/defer_spec.ts (hunk 4 above)
Comment (AndrewKushnir) [reply packages/core/test/acceptance/defer_spec.ts:1235]: Would it be possible to repro this in `main` using `afterRender` and a similar component structure? It'd be interesting to investigate to make sure there are no issues with `afterRender` timing.

**Review by AndrewKushnir — COMMENTED** [review]
@crisbeto looks great! I've left a few comments. I think we'd need to update how we keep track of prefetching triggers before we merge the PR.

File: packages/core/src/render3/instructions/defer.ts [hunk 11]
```diff
@@ -234,17 +236,36 @@ export function ɵɵdeferPrefetchOnHover() {}  // TODO: implement runtime logic.
...
  */
-export function ɵɵdeferPrefetchOnInteraction(target?: unknown) {}  // TODO: implement runtime logic.
+export function ɵɵdeferPrefetchOnInteraction(triggerIndex: number, walkUpTimes?: number) {
+  const lView = getLView();
+  const tNode = getCurrentTNode()!;
+  const tView = lView[TVIEW];
+  const tDetails = getTDeferBlockDetails(tView, tNode);
+
+  if (tDetails.loadingState === DeferDependenciesLoadingState.NOT_STARTED) {
+    tDetails.loadingState = DeferDependenciesLoadingState.SCHEDULED;
```
[the remark this answers is not in the document]
Comment (crisbeto) [reply packages/core/src/render3/instructions/defer.ts:264]: I'm not sure I understand what the issue is here. Aren't both triggers trying to start the preloading of the same resources? Hence we care until one of them fires first after which we can cancel the rest of the triggers. We can move this check inside the `afterRender` to avoid binding any events if `on idle` started it already.

File: packages/core/src/render3/instructions/defer_event_manager.ts (hunk 9 above)
Comment (crisbeto) [reply packages/core/src/render3/instructions/defer_event_manager.ts:29]: Yes, I was planning on having two more methods for `hover` and `viewport`, but I actually ended up with having one method. I can also rework it so it's a few top-level function, I just need to verify that we'd be able to inject `NgZone` for `on viewport`.

File: packages/core/src/render3/instructions/defer_event_manager.ts (hunk 10 above)
Comment (crisbeto) [reply packages/core/src/render3/instructions/defer_event_manager.ts:62]: I'll give it a try, but my intuition was that it wouldn't be worth the hassle since it's just two events and the `addEventListener` and `removeEventListener` calls are right next to each other so it'll be easy to see that we need to update two places if we were to add another event.

File: packages/core/test/acceptance/defer_spec.ts (hunk 4 above)
Comment (crisbeto) [reply packages/core/test/acceptance/defer_spec.ts:1235]: From my investigation, `afterRender` was definitely firing. The problem was that it's asynchronous and we need it to fire before the relevant assertion. I'm not sure that the `defer` tests should be concerned with the `afterRender` timing. That's why `afterRender` has tests of its own.

File: packages/core/src/render3/instructions/defer.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (AndrewKushnir) [reply packages/core/src/render3/instructions/defer.ts:264]: This logic schedules resource loading operation, but doesn't actually initiate the loading process. As a result, there might be multiple prefetch triggers that would result in scheduling multiple operations. Once the first one is triggered, we should cleanup the rest. In some cases (e.g. `on idle` or when we do `on interaction(localRefFromTemplate)`), we do not want to schedule prefetching more than once per defer block type (that may have multiple instances, for ex. if a defer is in a for loop).

I've put together a PR with a proposed change to unify the management of prefetch triggers and cleanup functions: https://github.com/angular/angular/pull/51856. Please let me know what you think about it.

**Comment by crisbeto** [comment]
I've addressed all of the feedback, except the one that depends on https://github.com/angular/angular/pull/51856. I also ported over some of the changes from my `on hover` and `on viewport` branch.

**Review by AndrewKushnir — APPROVED** [review]
👍

**Comment by dylhunn** [comment]
This PR was merged into the repository by commit 3cbb2a8ecf202c814c37ab241ce9f57fb574692e.
